test/03_device-state.ts

```diff
@@ -510,6 +555,316 @@ export default () => {
+										while (Date.now() < keepPollingForUntil) {
+											await fakeDevice.getState(
+												device3,
+												device3.uuid,
+												stateVersion,
+											);
+											await setTimeout(maxGetStateEventConsumptionTimeout);
+											// B/c the write cache's state was already online, there was no device PATCH
+											expect(device3ChangeEventSpy.called).to.be.false;
+											await expectDeviceHeartbeat(device3.id, {
+												db: DeviceOnlineStates.Offline,
+												cache: DeviceOnlineStates.Online,
+											});
+
+											console.log(`⌚  Waiting for next state GET poll...`);
+											await setTimeout(POLL_MSEC);
+										}
+
+										expect(
+											await getHeartbeatWriteCacheState(device3.id),
+										).to.have.property('currentState', 'online');
+										await expectResourceToMatch(
+											pineUser,
+											'device',
+											device3.id,
+											{
+												// We use the same description as in the itExpectsError, so that we are sure which part failed.
+												api_heartbeat_state: (prop, value) =>
+													prop.to.equal(
+														DeviceOnlineStates.Online,
+														`The api_heartbeat_state should have recovered to online but it was found ${value}`,
+													),
+											},
+										);
+									},
+									/The api_heartbeat_state should have recovered to online but it was found offline/,
+								);
+							});
+
+							describe('when the write cache expires while the device is Online (eg API stops serving requests)', function () {
+								let device3: fakeDevice.Device;
+								const device3ChangeEventSpy = sinon.spy();
+
+								before(async function () {
+									device3 = await fakeDevice.provisionDevice(
+										admin,
+										applicationId,
+									);
+									stateMock.getInstance().on('change', (args) => {
+										if (device3.id === args.deviceId) {
+											device3ChangeEventSpy(args);
+										}
+									});
+									await fakeDevice.getState(
+										device3,
+										device3.uuid,
+										stateVersion,
+									);
+									await waitFor({
+										checkFn: () => device3ChangeEventSpy.called,
+									});
+
+									await expectDeviceHeartbeat(device3.id, {
+										db: DeviceOnlineStates.Online,
+										cache: DeviceOnlineStates.Online,
+									});
+									// emulate the write cache having expired while the redis message queue hasn't been processed
+									// eg b/c the API/DB was down.
+									await redis.del(getHeartbeatWriteCacheKey(device3.id));
+								});
+
+								beforeEach(function () {
+									device3ChangeEventSpy.resetHistory();
+									delete tracker.states[device3.id];
+								});
+
+								itExpectsError(
+									'should stay Online and not consume the stale heartbeat redis queue message when a newer state GET has arrived',
+									async function () {
+										const initialWait = 0.8 * devicePollInterval;
```

Comment:
Such accurate wait requirements makes me worry about test flakiness. Is that an invalid concern?

Reply:
It shouldn't, except if we make the mock poll interval even smaller.
Let me leave a comment on how the number was picked.

Reply:
I also added bigger descriptions of what the two describes are trying to test:

// Here we test how the heartbeat behaves when the write cache has expired (eg b/c of downtime) and when the API starts working again,
// the first state GET request that we receive for a device happens while its heartbeat has switched to Timeout
// (the stale redis queued message to switch to Timeout has been consumed, and one to go Offline has been scheduled).

// Here we test how the heartbeat behaves when the write cache has expired (eg b/c of downtime) and when the API starts working again,
// the first state GET request that we receive for a device happens while its heartbeat is still Online
// (the stale redis queued message to switch to Timeout has not yet be consumed).